sw/device/lib/dif/dif_aes.c

```diff
@@ -300,6 +317,87 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
   return kDifOk;
 }
 
+dif_result_t dif_aes_load_aad(const dif_aes_t *aes, const dif_aes_aad_t aad,
+                              bool set_aad_mode) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (set_aad_mode) {
+    // Configure valid AAD bytes.
+    uint32_t reg = bitfield_field32_write(
+        0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, aad.valid_bytes);
+    // Put AES-GCM into AAD mode.
+    reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
+                                 AES_CTRL_GCM_SHADOWED_PHASE_VALUE_GCM_AAD);
+    aes_shadowed_write(aes->base_addr, AES_CTRL_GCM_SHADOWED_REG_OFFSET, reg);
+  }
+
+  if (!aes_input_ready(aes)) {
+    return kDifUnavailable;
+  }
+
+  aes_set_multireg(aes, &aad.data[0], AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_set_gcm_text_mode(const dif_aes_t *aes,
+                                       size_t valid_bytes) {
+  if (aes == NULL) {
+    return kDifBadArg;
+  }
+
+  if (!aes_idle(aes)) {
+    return kDifUnavailable;
+  }
+
+  // Configure valid TEXT bytes.
+  uint32_t reg = bitfield_field32_write(
+      0, AES_CTRL_GCM_SHADOWED_NUM_VALID_BYTES_FIELD, valid_bytes);
+  // Put AES-GCM into TEXT mode.
+  reg = bitfield_field32_write(reg, AES_CTRL_GCM_SHADOWED_PHASE_FIELD,
+                               AES_CTRL_GCM_SHADOWED_PHASE_VALUE_GCM_TEXT);
+  aes_shadowed_write(aes->base_addr, AES_CTRL_GCM_SHADOWED_REG_OFFSET, reg);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_set_gcm_tag_mode(const dif_aes_t *aes, uint64_t len_ptx,
```

Comment:
Nit: this function doesn't just set the tag phase (not mode) but it actually computes the tag.